custom channels: use `tapd` as the aux component implementation for `lnd` in integrated mode

guggero opened the pull request.

ellemouton commented on config.go:
```diff
@@ -395,6 +397,17 @@ func loadAndValidateConfig(interceptor signal.Interceptor) (*Config, error) {
+	// custom error messages to peers through the SendCustomMessage RPC in
+	// lnd. Since the error messages aren't in the custom range, we
+	// explicitly need to allow them.
+	if !cfg.lndRemote {
```
ok so for remote mode i assume (like rpc middle ware stuff) that we will just start up but then quickly tap will notice it fails to deliver custom msgs to lnd?

ellemouton commented on terminal.go:
```diff
@@ -1233,6 +1248,46 @@ func (g *LightningTerminal) BuildWalletConfig(ctx context.Context,
+	tapdWrapper, available := g.subServerMgr.GetServer(subservers.TAP)
+	if !available {
+		return nil, fmt.Errorf("tapd is not available, must be " +
+			"started in integrated mode for Taproot Assets " +
```
`started in integrated mode` -> do you mean that tapd must be integrated? cause 'integrated mode' to me means that lnd is integrated

guggero replied:
Actually, both. I'll update the error message.

ellemouton commented on terminal.go:
```diff
@@ -818,16 +818,19 @@ func (g *LightningTerminal) setUpLNDClients(lndQuit chan struct{}) error {
 				BlockUntilUnlocked:    true,
 				CallerCtx:             ctxc,
 				CheckVersion:          minimalCompatibleVersion,
+				ChainSyncPollInterval: g.cfg.LndConnectInterval,
```
nice

guggero replied:
Yeah, should speed up the normal itest as well.

ellemouton commented on cmd/litcli/ln.go:
```diff
@@ -0,0 +1,751 @@
+
+var lnCommands = []cli.Command{
+	{
+		Name:     "ln",
```
would `tap` be better? `ln` to me is general `lightning network`

guggero replied:
Yeah, I agree. Though I think this is temporary and we'll remove them in a future release anyway. But for now I think Polar already uses the commands in some demos, so I don't want to rename them just for renaming sake.

dstadulis replied:
I suppose since it's inducing ln functionality but pairing with tapd it's a bit of both

ellemouton commented on cmd/litcli/ln.go:
```diff
@@ -0,0 +1,751 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/record"
+	"github.com/urfave/cli"
+)
```
just skimmed this commit really quickly but it isnt yet immediately obvious to me why this is here and not in the taproot assets `cmd` directory? ie what makes this different from `loopcli` or equivalent?

guggero replied:
We do eventually want the logic that's done in this file be RPCs in `tapd`, then it will make sense to move them to `tapcli`. But for now we need access to both `lnd`'s and `tapd`'s RPCs from within a single command, which is done most conveniently in `tapcli` (see the way we bake a macaroon just to connect to `tapd`).

ellemouton reviewed (commented):
Had a quick look to gain familiarity - looks good and mostly plumbing 👍  main question is around why the cli commands belong here

guggero commented on config.go:
```diff
@@ -379,6 +380,7 @@ func loadAndValidateConfig(interceptor signal.Interceptor) (*Config, error) {
 			cfg.Remote.LitDebugLevel, cfg.Lnd.LogWriter,
 		)
 	} else {
+		SetupLoggers(cfg.Lnd.LogWriter, interceptor)
```
Yes, you're right. I was debugging something logging related and forgot to remove this again. Confirmed it's not needed.

bitromortac commented on terminal.go:
```diff
@@ -466,13 +474,20 @@ func (g *LightningTerminal) start() error {
 			DatabaseBuilder:     g.defaultImplCfg.DatabaseBuilder,
 			WalletConfigBuilder: g,
 			ChainControlBuilder: g.defaultImplCfg.ChainControlBuilder,
+			AuxComponents:       *auxComponents,
```
Very cool to see how things come together :tada:!

bitromortac reviewed (commented):
Really exciting to see how everything fits together :tada::rocket:. Left some comments in this first (quick) pass.

ellemouton commented on cmd/litcli/main.go:
```diff
@@ -244,3 +295,69 @@ func printRespJSON(resp proto.Message) { // nolint
+
+	defer cleanup()
+
+	var suffixBytes [4]byte
+	superMacID := session.NewSuperMacaroonRootKeyID(suffixBytes)
+	lndClient := lnrpc.NewLightningClient(lndConn)
+	ctxb := context.Background()
```
personally, I like when these types of resources are defined at the top of a function (so i think this location is good).

guggero commented on terminal.go:
```diff
@@ -1233,6 +1248,46 @@ func (g *LightningTerminal) BuildWalletConfig(ctx context.Context,
+	if tapdWrapper.Remote() {
+		return nil, fmt.Errorf("tapd is not available, must be " +
+			"started in integrated mode for Taproot Assets " +
+			"Channels to be available")
```
Updated the error message according to @ellemouton's review. Hope it makes more sense now.

guggero commented on cmd/litcli/ln.go:
```diff
@@ -0,0 +1,751 @@
+	msatPerUnit := acceptedQuote.BidPrice
+	numUnits := uint64(decodeResp.NumMsat) / msatPerUnit
+
+	fmt.Printf("Got quote for %v asset units at %v msat/unit from peer "+
```
Hmm, this message isn't really about consent, more of an information (and I guess initially mainly added for debugging purposes :sweat_smile: ).
The actual validation of the exchange rate is done by the RFQ and price oracle logic of `tapd`. The end user will be able to plug in their exchange API of choice to validate prices against, and then define a maximum percentage (PPM actually) of deviation due to spread.

guggero commented on cmd/litcli/ln.go:
```diff
@@ -0,0 +1,751 @@
+		ValueMsat:       int64(numMSats),
+		DescriptionHash: descHash,
+		FallbackAddr:    ctx.String("fallback_addr"),
+		Expiry:          ctx.Int64("expiry"),
```
Ah yes, good point. Fixed!

guggero commented on itest/litd_custom_channels_test.go:
```diff
@@ -0,0 +1,1339 @@
+	fabiaAssetBalance += fabiaInvoiceAssetAmount1
+
+	// ------------
+	// Test case 4: Create an asset invoice on Fabian and pay it with just
```
It's a bit hidden, we hard code the max shard size to 80k in the `payInvoiceWithXXX()` helper. So choosing a large enough asset amount will guarantee at least two shards. Updated the comment to make it a bit more obvious.
And yes, we don't currently track the number of attempts. There are a lot of assertions still missing here, mainly due to the fact that the itest library isn't fully compatible with the refactored version in `lnd`, so we can't just re-use a lot of those assertions.

guggero commented:
Thanks for the reviews! I addressed all comments and fixed the failing CI steps (hopefully).
Taking the PR out of draft, even if we still want to add the force-close code. But that will likely only affect the integration test (and the `go.mod` file).
